arch/x86/pagetables.c

```diff
@@ -537,6 +620,235 @@ frame_t *find_user_va_frame(const void *va) {
     return find_va_frame(&user_cr3, va);
 }
 
+static inline void *_vmap_range_chunk(cr3_t *cr3_ptr, void *va, mfn_t mfn,
+                                      unsigned int order, unsigned long flags,
+                                      bool propagate_user) {
+    switch (order) {
+    case PAGE_ORDER_4K:
+        return _vmap_4k(cr3_ptr, va, mfn, flags, propagate_user);
+    case PAGE_ORDER_2M:
+        return _vmap_2m(cr3_ptr, va, mfn, flags, propagate_user);
+    case PAGE_ORDER_1G:
+        return _vmap_1g(cr3_ptr, va, mfn, flags, propagate_user);
+    default:
+        BUG();
+    }
+
+    return NULL;
+}
+
+static inline int _vmap_range(mfn_t mfn, unsigned int order, unsigned long flags,
+                              vmap_flags_t vmap_flags) {
+    const int err = -EFAULT;
+
+    if (vmap_flags & VMAP_KERNEL) {
+        if (!_vmap_range_chunk(&cr3, mfn_to_virt_kern(mfn), mfn, order, flags, false))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_IDENT) {
+        if (virt_invalid(
+                _vmap_range_chunk(&cr3, mfn_to_virt(mfn), mfn, order, flags, false)))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_KERNEL_MAP) {
+        if (!_vmap_range_chunk(&cr3, mfn_to_virt_map(mfn), mfn, order, flags, false))
+            return err;
+    }
+
+    if (vmap_flags & (VMAP_KERNEL_USER | VMAP_KERNEL_USER_ACCESS)) {
+        unsigned long _flags = flags & ~_PAGE_USER;
+        if (vmap_flags & VMAP_KERNEL_USER_ACCESS)
+            _flags |= _PAGE_USER;
+
+        if (!_vmap_range_chunk(&cr3, mfn_to_virt_user(mfn), mfn, order, _flags, false))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_USER) {
+        if (!_vmap_range_chunk(&user_cr3, mfn_to_virt_user(mfn), mfn, order,
+                               flags | _PAGE_USER, true))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_USER_IDENT) {
+        if (!_vmap_range_chunk(&user_cr3, mfn_to_virt(mfn), mfn, order,
+                               flags & ~_PAGE_USER, false))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_USER_KERNEL) {
+        if (!_vmap_range_chunk(&user_cr3, mfn_to_virt_kern(mfn), mfn, order,
+                               flags & ~_PAGE_USER, false))
+            return err;
+    }
+
+    if (vmap_flags & VMAP_USER_KERNEL_MAP) {
+        if (!_vmap_range_chunk(&user_cr3, mfn_to_virt_map(mfn), mfn, order,
+                               flags & ~_PAGE_USER, false))
+            return err;
+    }
+
```

Comment:
I was thinking about it too, but decided it is too much work for very little gain as the errors are (supposed to be) very rare and typically fatal. So I decided to return an error and leave the ultimate decision to the caller what to do. 
But, I will add a comment saying this might be needed in the future.